WIP: Changes needed for NAG Compiler

mathomp4 opened the pull request:
This PR tracks changes needed to compile GMAO_Shared with the NAG Compiler. NAG is *very* strict and will not allow the use of Fortran extensions that are endemic to GEOS. 

---

The first category are procedures that are now part of the Fortran Standard that used to be extensions in some compilers:

- `iargc()` → `command_argument_count()`
- `getarg()` → `get_command_argument()`
- `getenv()` → `get_environment_variable()`
- `system()` → `execute_command_line()`
- `call exit(0)` → `stop`
- `call exit(N)` → `error stop N`

NAG also doesn't support (without flags that need to be used *everywhere* as @tclune can elucidate on) the use of `real*4` or `real(kind=4)`. The reason is that NAG does not use 4 for 32-bit real and 8 for 64-bit, instead they use 1 and 2. (The Standard does not specify what "kind" a 32-bit real must be, only that there must be one.) So, many changes are of the type:

- `real*4` → `real(REAL32)`
- `real(kind=8)` → `real(kind=REAL64)`

mathomp4 committed:
Changes needed by NAG

mathomp4 committed:
Merge branch 'main' into feature/mathomp4/nag-fixes

mathomp4 committed:
More NAG updates

mathomp4 committed:
Merge branch 'main' into feature/mathomp4/nag-fixes

mathomp4 committed:
More NAG updates

mathomp4 committed:
Fix for rs_numtiles

mathomp4 committed:
Remove unneeded contains

mathomp4 committed:
Merge branch 'main' into feature/mathomp4/nag-fixes

mathomp4 committed:
Merge branch 'main' into feature/mathomp4/nag-fixes

rtodling reviewed (dismissed):
These are fine by me. Branch might need update though.

mathomp4 commented:
> These are fine by me. Branch might need update though.

Yeah. I need to revisit this. Files got moved long ago...

I'll put it on the list!

weiyuan-jiang committed:
Fix NAG compiler issues in GMAO_Shared

- Fix m_ioutil.F90: Use TRANSFER() intrinsic for type-safe byte swapping
  in swapI4_ and swapI8_ functions instead of direct type punning

- Fix mpi0/mpi0_copy.F90: Implement TRANSFER-based type conversions using
  BLOCK constructs for all non-INTEGER MPI types (REAL, DOUBLE_PRECISION,
  LOGICAL, CHARACTER) to satisfy NAG's strict type checking

- Fix windfix.F90: Correct array section syntax from scalar notation
  (e.g., dglo(1,1,l)) to proper array slices (e.g., dglo(:,:,l)) for
  calls to GETDIV and VELPOT_SP

- Update GMAO_mpeu/CMakeLists.txt: Comment out duplicate MISMATCH flags
  that are now in common_Fortran_flags to avoid compilation errors

All changes maintain backwards compatibility with gfortran and other
compilers while fixing NAG compiler strict type checking issues.

Commit:
Merge branch 'main' into feature/mathomp4/nag-fixes

Commit:
Fix NAG compiler issues in GMAO_gfio and GMAO_hermes

- GMAO_gfio/CMakeLists.txt: add -wmismatch for legacy NetCDF-2 C API
  routines (NCVDEF, NCAPT, NCVPT, NCAGT, NCVGT, NCVGT1) called with
  varying Fortran types intentionally
- GMAO_hermes/CMakeLists.txt: remove duplicate ${MISMATCH} from
  CMAKE_Fortran_FLAGS_RELEASE (already in common_Fortran_flags); add
  -wmismatch for GFIO_PUTREALATT/GFIO_GETREALATT type-punning
- GMAO_hermes/m_nc_JEDIinc.f90: replace byte-count kind literal real(4)
  with portable real(selected_real_kind(6)) via parameter r4
- GMAO_hermes/m_nc_akbk.f90: replace double precision with real(real64)
  from iso_fortran_env so NF90_PUT_VAR generic resolves correctly
  when compiled with -r8

Commit:
changed by Claude

weiyuan-jiang commented on CMakeLists.txt:
```diff
@@ -11,7 +11,7 @@ if ( NOT HERMES_LIGHT )
      GMAO_pilgrim
      GMAO_etc
-     GEOS_Util
+     @GEOS_Util
```
It seems GEOS_Util has been deleted. But this @GEOS_Util should be here?
